### DISCUSSION TITLE
Global Scene Network Objects - Hear Me Out Please.

### DISCUSSION THREAD

**Discussion by Katerlad** [FirstGearGames/FishNet#488] [Feature Request]
I know this has been brought up many times in the past, but I feel an in depth conversation has not been had to find a solution. If so I wasn't there to start a fire 🔥 😆 

# Background
I remember one of your last conclusions on this was that Scene Network Objects cannot be marked global because the clients would of had to at least load that scene once to be able to also spawn the object and be synced together. Which is not ideal, so you have disabled this functionality as a protection to the user to not make this mistake.

# Potential Solution
However! Correct me if I am wrong, but is there not a way to just require Global Scene Network Objects to be prefabs the DefaultPrefabsObject has registered?, and upon the client connection to the server, the server will look for any global objects that were scene objects and Spawn them as global to the clients? 

I understand that in the Scenes you would not be able to modify the Scene Object Prefab to have Custom Values in any of the components that are not Network Behaviours, because the server would only ever be able to spawn the DefaultPrefabObject. 

^ I feel that putting this in the docs as what the restriction actually is, and to allow users to still drag default prefab objects as a Scene Object and mark them global is the way to go!

# Conclusion
I get why a normal scene object that is not a prefab, cannot be handled this way, because the server has no Idea what object to spawn., but as long as the scene object is a prefab. The scene object should have an NetworkID associated and registered in the defaultPrefabObjects, and the server should be able to spawn this in for clients.

Im probably missing something silly that makes this all moot point, but I don't want to go down without a fight haha.

When you have time to even discuss this which may not be awhile, we should get together on a discord call and hash out the details!

**Comment by FirstGearGames** [comment]
You actually can mark the prefab as global but not any instances placed in the scene. In fact, if you have a scene object marked as global FN warns you that the scene object will not be set global.

I do not think it's worth adding the complexity of automatically spawning scene prefabs just to make them global. Users could write a simple script and slap it on the NetworkManager to instantiate any number of prefabs as global upon start. Or the script could go in the scene.

Going to have to decline this one, sorry.

**Reply by Katerlad** [reply]
While I understand there is a lot of complexity, there is something to be said for keeping the workflow of some things as close to Unitys vanilla workflow as possible. 

I know a lot of people learn from tutorials and training guides to use Bootloader/Mainmenu scenes to drag your Manager Gameobjects in and be able to visualize them all in your bootloader Scene. 

They are also still allowed to do this in Mirror as well, and was one of my biggest gripes coming from Mirror to fishnet awhile back. 

Like you've mentioned in the past though even doing it with Mirror if you don't do it correctly you'll run into lots of problems, but if you do do it correctly ie. making sure all clients load that scene at least once before putting them in the other scenes. You can still use this Architecture.

So in the end, Mirror lets you do this, and lets the Dev decide your fate if you accidently implement incorrectly. 

Fishnet approach is to limit this functionality, to protect less experienced Devs from making a mistake. 

Im not sure if I agree with Fishnets approach, but also maybe its for the best.. Its an internal struggle, and I feel for anyone else who may come from mirror or NGO who run into this and may quietly go back because it was different to a way they are use to.

**Reply by FirstGearGames** [reply]
The main thing I'm seeing is you are not really describing 'global scene network objects', because this is 100% not possible without always having the scene loaded.

This is no different than using a bootloader scene in an offline game. Presumably when using bootloaders you are marking your objects as DDOL. You can do the same with FishNet simply by telling the server which prefabs to spawn, and by marking those prefabs as global in the inspector or at runtime prior to spawning.

I don't think I'm understanding the limitation, if anything it should be easier in FishNet because you can simply just instantiate on the server as global from anywhere and you're good.

**Reply by Katerlad** [reply]
Its just a workflow preference thing, a lot of people probably use instantiation of their prefabs with a DDOL script but a lot just like dragging their manager objects into the boot loader scene/starting scene and visualizing them in their hierarchy. 

As you see a lot in your discord server, many people just get confused that they can't do their normal workflow. Ie. dragging the objects into the boot load scene and putting a DDOL script on them. 

In the end it's just a small learning curve for those types of people, but if it was something that could be added into fishnet in a day, people would probably appreciate they can use their normal workflow.

**Reply by Katerlad** [reply]
Sounds like though the work would take too long to justify the benefit, but as long as it has been thought about in this context is all that matters to me! I will respect any final decision you make.

**Reply by nasearle** [reply]
How are you meant to load global manager objects so that they're guaranteed to be available when the client starts? The issue I'm running into with spawning the manager objects instead of simply marking them as DontDestroyOnLoad is that the client doesn't know when the object exists. You either have to create a separate scene to load the manager objects and only after they've loaded move on to the scenes where they'll be used, or you have to wait for a spawned event in every script that uses the manager objects. Being able to just mark these objects as DontDestroyOnLoad like in NGO would really simplify things.

**Comment by fdfer06** [comment]
zThe problem is that when having an object with more than 50 variables in the scene, such as those of a canvas, for each text it is complicated to have to assign everything from code, since it is something that is very simple to drag those objects from the inspector, being a very long task (not complicated) to do.

<img width="1889" height="1077" alt="image" src="https://github.com/user-attachments/assets/cf891005-bf66-4c18-a2a0-dc336d93b362" />
